hare/eligibility/oracle.go

```diff
@@ -245,6 +253,7 @@ func (o *Oracle) actives(layer types.LayerID) (map[string]struct{}, error) {
 		o.Error("Could not calculate hare active set size: no contextually valid blocks",
 			log.Uint64("layer_id", uint64(layer)), log.Uint64("epoch_id", uint64(layer.GetEpoch(o.layersPerEpoch))),
 			log.Uint64("safe_layer_id", uint64(sl)), log.Uint64("safe_epoch_id", uint64(safeEp)))
+		o.lock.Unlock()
 		return nil, errNoContextualBlocks
 	}
 
```

Comment:
Does mp gives the same value over different safe layers ? if so maybe this is wrong. you save the value by epoch but this calculation is done by safe later.

Reply:
also if getActiveSet can start calculation for next layer/epoch already so we need to prallelize a little bit.

Reply:
Yes but it should be a relatively quick query.
I added that optimization.